### PR TITLE
#9578: Add WITH_PYTHON_BINDINGS option to allow build w/o python

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/9578

### Problem description
We expect that many customers focus on C++ part of ttnn, so we want to provide an easy way to build without python bindings. 

### What's changed
Adding a new build flag `WITH_PYTHON_BINDINGS` which is ON by default.
When `OFF`, we do not look for Python3 and pybind, do not create pybind_pch, do not create tt_lib target and exclude bindings from ttnn lib.

I expect that that consumers will link `ttnn` like this
```
add_target_library(my_app PRIVATE/PUBLIC ttnn)
```

🗓️ I think this solution is good for the next couple months.
The suggestion in the ticket was to follow the tt_dnn/tt_lib pattern, where the C++ OP library is separated from the lib with Py bindings. While in this PR I do not follow the suggestion, **I actually like 😊** how 
- tt_eager provides core infra, 
- tt_dnn provides ops and kernels and 
- tt_lib provides bindings. 

We will further discuss with the team on how to move long term.

### Checklist
- [ ] Post commit CI passes
- [x] Existing gtests cover this usecase

### CODE REVIEW

**Comment by ayerofieiev-tt** [comment]
@marty1885 please take a look, share your thoughts

File: tt_eager/tt_lib/CMakeLists.txt
Comment (marty1885) [inline tt_eager/tt_lib/CMakeLists.txt:4]: I think you meant `if(NOT WITH_PYTHON_BINDINGS)`? Else it errors on my dev machine (CMake 3.29).

```
CMake Error at tt_eager/tt_lib/CMakeLists.txt:2 (if):
  if given arguments:

    "not" "WITH_PYTHON_BINDINGS"

  Unknown arguments specified
```

**Review by marty1885 — COMMENTED** [review]
I think this will work in the short term. Me too. I like it when the libraries are separated. Hope TT can get to that point soon (and reenable stdlibc++). in the mean time I can hack my way around the library borders.

Long term I think it will be a pain as users have to compile metalium twice and deal with 2 shared objects but with different linkage. 

This branch errors for me when compiling with error. I think this is CMake trying to build TTNN without Python enabled?

```
/home/marty/Documents/not-my-projects/tt-metal/tt_metal/third_party/pybind11/include/pybind11/detail/../detail/common.h:266:10: fatal error: 'Python.h' file not found
  266 | #include <Python.h>
      |          ^~~~~~~~~~
1 error generated.
```

If you ever got around to implement #9574. Please expose this linkage information to the exported CMake target.

**Review by tt-rkim — APPROVED** [review]

**Review by TT-billteng — APPROVED** [review]

**Review by vtangTT — APPROVED** [review]

**Comment by tt-rkim** [comment]
Sorry to say this after it's merged... but should we expose through `build_metal.sh`, or leave it as an undocumented feature?

**Comment by marty1885** [comment]
Your call. I'm ok both ways since this (supposedly) is not a long term fix. Hopefully the proper solution gets implemented before I start upstreaming my GGML backend. I'm ok with this being undocumented for my local development. If you feel there will be other consumers of this feature, maybe document it?

**Pull request merged**
